prov/sm2/src/sm2_ep.c

```diff
@@ -297,6 +297,49 @@ static ssize_t sm2_do_inject(struct sm2_ep *ep, struct sm2_region *peer_smr,
 	return FI_SUCCESS;
 }
 
+static void sm2_format_cma(struct sm2_xfer_entry *xfer_entry,
+			   enum fi_hmem_iface iface, uint64_t device,
+			   const struct iovec *iov, size_t count)
+{
+	struct sm2_cma_data *cma_data =
+		(struct sm2_cma_data *) xfer_entry->user_data;
```

Comment:
semantically I think it makes more sense to turn this into a union now instead of casting it

Reply:
done.

Reply:
Now that I have made user_data[] a flexible array, I need to undo this change b/c you can't have flexible arrays in a union:

```
In file included from prov/sm2/src/sm2_cntr.c:34:
prov/sm2/src/sm2.h:136:25: error: flexible array member in union
  136 |                 uint8_t user_data[];
```